metrics/internal/collectors/storage-cluster.go

```diff
@@ -0,0 +1,88 @@
+package collectors
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+
+	v1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	"github.com/red-hat-storage/ocs-operator/metrics/internal/options"
+)
+
+type StorageClusterCollector struct {
+	KMSServerConnectionStatus *prometheus.Desc
+	Informer                  cache.SharedIndexInformer
+}
+
+var _ prometheus.Collector = &StorageClusterCollector{}
+
+func NewStorageClusterCollector(opts *options.Options) *StorageClusterCollector {
+	cl, err := GetOcsClient(opts)
+	if err != nil {
+		klog.Errorf("Unable to get client: %v", err)
+		return nil
+	}
+	lw := cache.NewListWatchFromClient(cl, "storageclusters", metav1.NamespaceAll, fields.Everything())
+	sharedIndexInformer := cache.NewSharedIndexInformer(lw, &v1.StorageCluster{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	return &StorageClusterCollector{
+		KMSServerConnectionStatus: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "storagecluster", "kms_connection_status"),
+			`KMS Connection Status; 0: Connected, 1: NotConnected, 2: Failure`,
+			[]string{"name", "namespace"},
+			nil,
+		),
+		Informer: sharedIndexInformer,
+	}
+}
+
+func (c *StorageClusterCollector) Run(stopCh <-chan struct{}) {
+	go c.Informer.Run(stopCh)
+}
+
+func (c *StorageClusterCollector) Describe(ch chan<- *prometheus.Desc) {
+	ds := []*prometheus.Desc{
+		c.KMSServerConnectionStatus,
+	}
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+func (c *StorageClusterCollector) Collect(ch chan<- prometheus.Metric) {
+	scLister := NewStorageClusterLister(c.Informer.GetIndexer())
+	storageClusters := getAllStorageClusters(scLister)
+	if len(storageClusters) > 0 {
+		c.collectKMSConnectionStatuses(ch, storageClusters)
+	}
+}
+
+func getAllStorageClusters(lister StorageClusterLister) []*v1.StorageCluster {
+	storageClusters, err := lister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("couldn't list StorageCluster: %v", err)
+		return nil
+	}
+	return storageClusters
+}
+
+func (c *StorageClusterCollector) collectKMSConnectionStatuses(ch chan<- prometheus.Metric, storageClusters []*v1.StorageCluster) {
+	for _, storageCluster := range storageClusters {
+		v := 1
+		if storageCluster.Status.KMSServerConnection.KMSServerAddress != "" {
+			v = 0
+			if storageCluster.Status.KMSServerConnection.KMSServerConnectionError != "" {
+				v = 2
+			}
+		}
```

Comment:
According to the above description (line#33), 0: Connected , 1: Not Connected and 2: Failure
But what is the distinction between `Not Connected` and a `Failure` (failure means it is not connected, right?)
IMHO, we can have THREE states, `Connected`, `Failure` or `Not Connected` and `KMS not enabled`.

Consider this scenario where,
1. `KMSServerAddress` is empty because of a connection failure, that is we have an error message in `KMSServerConnectionError` field.

But as per the above conditions, we will never reach `v = 2` (Failure scenario) and always be in not connected state (which is technically correct) =)

Reply:
This use case is indeed better, but it seems like we have some limitations as to how code is working as of now.
In the code [here](https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/cephcluster.go#L177), we are updating the `KMSServerAddress`, `KMSServerConnectionError` only when the provider is `vault` and the configmap data field has the `VAULT_ADDR`, else we don't update the `KMSServer*` values and just report an error. This was added in the PR #1419, @umangachapagain do you know as to why we have limited the update only for this particular case? Shouldn't we keep it generic? Like, maybe not update the address for other providers but at least update the `KMSServerConnectionError` value? Else if the KMS connection fails for a provider other than `vault` we have no way to identify and generate an alert.

Reply:
So, I have updated the code to add address and error status in case of an error irrespective of the KMS provider. For any provider other than `vault` the address field will be empty but the error field will be updated with the respective error.